### PR TITLE
[Model][LoRA]LoRA support added for MolmoForCausalLM

### LINKED ISSUE #11431: [Feature]: AssertionError: MolmoForCausalLM does not support LoRA yet.
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
PyTorch version: 2.5.1+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-122-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.5.82
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA RTX A6000

Nvidia driver version: 555.42.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen Threadripper PRO 5955WX 16-Cores
CPU family:                           25
Model:                                8
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             2
Frequency boost:                      enabled
CPU max MHz:                          7031.2500
CPU min MHz:                          1800.0000
BogoMIPS:                             7985.02
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                       AMD-V
L1d cache:                            512 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             8 MiB (16 instances)
L3 cache:                             64 MiB (2 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Mitigation; safe RET
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-ml-py==12.560.30
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] pyzmq==26.2.0
[pip3] torch==2.5.1
[pip3] torchvision==0.20.1
[pip3] transformers==4.47.1
[pip3] triton==3.1.0
[conda] Could not collect
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.6.5
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
GPU0    GPU1    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      SYS     0-31    0               N/A
GPU1    SYS      X      0-31    0               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

CUDAPATH=/usr/local/cuda-12.5
CUDACXX=/usr/local/cuda-12.5/bin/nvcc
LD_LIBRARY_PATH=/molmo/.venv/lib/python3.10/site-packages/cv2/../../lib64:/usr/local/cuda-12.5/lib64:
CUDA_MODULE_LOADING=LAZY
```

</details>

### Model Input Dumps

_No response_

### 🐛 Describe the bug

Apparently MolmoForCasualLM does not yet support Lora adapters, yielding an AssertionError on serving:

`AssertionError: MolmoForCausalLM does not support LoRA yet.`

I trained a Lora adapter with HF Trainer and would like to use it together with vLLM for fast inference. This seems not implemented yet. I tested this by trying to serve Molmo directly via:
`vllm serve allenai/Molmo-7B-D-0924 --enable-lora --trust-remote-code --max-num-seqs 6 --tensor-parallel-size 1 --lora-modules test=$LORA_DIR/checkpoint-25`

Are there any plans to get this working or is there a guide somewhere how i can enable lora for Molmo myself?
If all works I'd be open to submit a PR but i'd need some guidance.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by ayylemao**
[Model][LoRA]LoRA support added for MolmoForCausalLM
LoRA supported for inference in MolmoForCausalLM

Need some help with verifying this works correctly since my own tests revealed the same results for base_model and base_model+ lora.

ping @jeejeelee 

FIX #11431

### CODE REVIEW

**Comment by ayylemao** [comment]
i might have done goofed by syncing my fork and pushing to the PR branch

File: vllm/model_executor/models/molmo.py [hunk 1]
```diff
@@ -1121,9 +1121,26 @@ def input_processor_for_molmo(ctx: InputContext, inputs: DecoderOnlyInputs):
...
+        "ff_proj",
+    ]
+    embedding_modules = {}
+    embedding_padding_modules = {}
+
+    def __init__(self,
+                 *,
+                 vllm_config: VllmConfig,
+                 prefix: str = "",
+                 lora_config: Optional[LoRAConfig] = None):
```
Comment (jeejeelee) [inline vllm/model_executor/models/molmo.py:1143]: We can obtain `lora_config` directly from `vllm_config` instead of adding it as a separate argument

**Comment by jeejeelee** [comment]
@ayylemao  could you plz address the branch conflict ?

**Comment by ayylemao** [comment]
Thank you for your review. 

I will get back to it after christmas!

File: vllm/model_executor/models/molmo.py (hunk 1 above)
Comment (ayylemao) [reply vllm/model_executor/models/molmo.py]: fixed that

**Comment by jeejeelee** [comment]
@ayylemao I've added modules with LoRA support. Could you test if it generates reasonable results?

**Comment by ayylemao** [comment]
> @ayylemao I've added modules with LoRA support. Could you test if it generates reasonable results?

After testing i still get the same results for the base model and base_model+lora when serving via vllm serve.
I think my lora adapter still does not get applied correctly, even though when evaluating via transformers+peft it works as expected.

I am not sure how to test more accuratly if the lora adapters get correctly applied or not beyond just comparing the actual model output.
Do you have any idea what might be the issue or how to pinpoint further?

**Comment by jeejeelee** [comment]
> > @ayylemao I've added modules with LoRA support. Could you test if it generates reasonable results?
> 
> @jeejeelee After testing i still get the same results for the base model and base_model+lora when serving via vllm serve. I think my lora adapter still does not get applied correctly, even though when evaluating via transformers+peft it works as expected.
> 
> I am not sure how to test more accuratly if the lora adapters get correctly applied or not beyond just comparing the actual model output. Do you have any idea what might be the issue or how to pinpoint further?

I'm training a LoRA model for this model and will test it tomorrow.

**Comment by jeejeelee** [comment]
> > > @ayylemao I've added modules with LoRA support. Could you test if it generates reasonable results?
> > 
> > 
> > @jeejeelee After testing i still get the same results for the base model and base_model+lora when serving via vllm serve. I think my lora adapter still does not get applied correctly, even though when evaluating via transformers+peft it works as expected.
> > I am not sure how to test more accuratly if the lora adapters get correctly applied or not beyond just comparing the actual model output. Do you have any idea what might be the issue or how to pinpoint further?
> 
> I'm training a LoRA model for this model and will test it tomorrow.


@ayylemao  After merging with https://github.com/vllm-project/vllm/pull/11551, when testing 100 samples, 99 generated results matched exactly with the `transformers` output. The remaining one also produced a reasonable result.

**Comment by ayylemao** [comment]
> > > > @ayylemao I've added modules with LoRA support. Could you test if it generates reasonable results?
> > > 
> > > 
> > > @jeejeelee After testing i still get the same results for the base model and base_model+lora when serving via vllm serve. I think my lora adapter still does not get applied correctly, even though when evaluating via transformers+peft it works as expected.
> > > I am not sure how to test more accuratly if the lora adapters get correctly applied or not beyond just comparing the actual model output. Do you have any idea what might be the issue or how to pinpoint further?
> > 
> > 
> > I'm training a LoRA model for this model and will test it tomorrow.
> 
> @ayylemao After merging with #11551, when testing 100 samples, 99 generated results matched exactly with the `transformers` output. The remaining one also produced a reasonable result.

Thank you for testing.
Could you share the lora modules you targeted? Or directly your Peft Config? It would help me a lot.

**Review by jeejeelee — APPROVED** [review]
@ayylemao  Thanks for your contribution

**Comment by jeejeelee** [comment]
@ayylemao  This is a known issue and is not related to this PR.

**Pull request merged**